### ISSUE TITLE
Add support, or recommend another library for dealing with already existing file formats using WTF-8.

### ISSUE THREAD

**Issue by Lymia** [SimonSapin/rust-wtf8#7]
Like it or not, there are already files out in the wild with WTF-8 encoding, before the specification was ever defined. The prime culprit of this that I'm aware of is Java's [`DataOutputStream.writeUTF`](https://docs.oracle.com/javase/7/docs/api/java/io/DataOutputStream.html#writeUTF(java.lang.String)) function which is very commonly used in Java applications, and outputs an encoding that (AFAIK) is identical to WTF-8.

The [`cesu8`](https://docs.rs/cesu8/1.1.0/cesu8/) crate is the next best option I'm aware of, but even that is not a perfect solution as Java `String` does not enforce that its contents are valid UTF-16, and they can contain unpaired surrogates that eventually end up in file formats.

tl;dr: Unfortunately, while the newer spec says this should not be done, this has been done since at least 1996.

**Comment by Lymia** [comment]
Ergh. Never mind this on the Java use case. I hadn't noticed that WTF-8 disallows paired surrogate pairs entirely, which means Java encodes something that is neither WTF-8 or CESU-8. ¯\\\_(ツ)_/¯

**Comment by SimonSapin** [comment]
Can you get away with `String::from_utf8_lossy` for those Java-generated files or do you need to preserved malformed content?

**Comment by Lymia** [comment]
I don't really need to preserve unpaired surrogates (yet), but I'd prefer to not just throw an error if they show up. AFAIK, `String::from_utf8_lossy` will eat paired surrogate pairs too, and `cesu8` will outright throw an error on unpaired surrogates.

That is probably more appropriate as a pull request to the `cesu8` crate, or just its own crate the more I look at it.
